rfc/rfc-0000-input-output-correlation.md

```diff
@@ -0,0 +1,223 @@
+resource that their templates stamp out.
+
+Example matching inputs against outputs with the `correlationRules` field:
+
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterSourceTemplate
+metadata:
+  name: example-source-scanning-thing
+spec:
+  template:
+    apiVersion: my-scanner/v1
+    kind: SourceScan
+    metadata:
+      name: ...
+    spec:
+      scanUrl: "$(sources.source.url)$"
+      sourceRevision: "$(sources.source.url)$"
+correlationRules:
+  - input: "$(sources.source.url)"              #evaluated against inputs in template context
+    output: "$(status.outputs.url)"             #evaluated against resource status
+  - input: "$(sources.source.revision)"         #evaluated against inputs in template context
+    output: "$(status.outputs.scannedRevision)" #evaluated against resource status
+```
+
+# How it Works
+
+[how-it-works]: #how-it-works
+
+When Cartographer processes a resource, the jsonpath expressions in each `correlationRules` `input` and `output` fields
+are evaluated and compared. If they are all equal, then the output is considered to be correlated with the input(s).
+
+When output cannot be correlated to input, Cartographer will not propagate the output forward. Doing so ensures we can
+always correlate outputs with inputs on subsequent resources, and also mitigates behavior of bad actors in the cluster
+as described
+in [Draft RFC Read Resources Only When In Success State](https://github.com/vmware-tanzu/cartographer/pull/556).
+
+In the absence of `correlationRules` Cartographer will continue to propagate outputs without attempting to correlate
+inputs to preserve existing functionality.
+
+For cases where the output does not contain necessary information to match against inputs, one may instead match against
+the spec of the resource:
+
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterImageTemplate
+metadata:
+  name: example-image-building-thing
+spec:
+  template:
+    apiVersion: my-builder/v1
+    kind: ImageBuilderThatDoesntExposeInputDetailsInItsOutput
+    metadata:
+      name: ...
+    spec:
+      sourceUrl: "$(sources.source.url)$"
+      sourceRevision: "$(sources.source.url)$"
+correlationRules:
+  - input: "$(sources.source.url)"              #evaluated against inputs in template context
+    output: "$(spec.sourceUrl)"                 #evaluated against resource spec
+  - input: "$(sources.source.revision)"         #evaluated against inputs in template context
+    output: "$(spec.sourceRevision)"            #evaluated against resource spec
+```
+
+Since the spec of the resource may change independently of the output, when the `correlationRules` makes any assertion
+against the `spec`, Cartographer must also ensure that the current output is a result of processing the current spec by
+making additional assertions:
+
+* Resource is healthy, as per [Allow Resources to Report Status](https://github.com/vmware-tanzu/cartographer/pull/738)
+* Resource `observedGeneration == generation`
```

Comment:
the template author should be required to indicate that the resource supports observedGeneration, many resources don't